### ISSUE TITLE
UX improvements for jacking in with additional aliases 

### ISSUE THREAD

**Issue by matt-y** [clojure-emacs/cider#3317]
**Is your feature request related to a problem? Please describe.**

Firstly this is an opinion that I hope is not taken in a negative way but as constructive feedback for cider. I'm a big fan of it. I appreciate your taking the time to read this. 

Loading an additional alias such as `:dev` or `:test` when jacking in is a very common operation that in its current state I feel is overly difficult. Current solutions involve installing additional emacs packages such as `dir-locals` and providing a per-project configuration that is used every time to load additional aliases at `jack-in` time.  Users are also able to use a prefix `C-u` when doing a jack-in to provide additional arguments: ``. This is not discoverable in the CIDER documentation on the web or inside emacs. If it indeed is documented, I apologize but I have failed at finding it.  I have only found reference to the prefix usage in the issue tracker here on github, or deep in the clojurians slack. I myself have asked how to load an alias there!

Additionally, after jacking in and getting a repl going, loading an alias requires the repl be stopped and re-started with that alias loaded explicitly. Some kind of porcelain in front of these manual steps would be nice. 

Some examples of users getting tripped up can be found below: 
- https://github.com/clojure-emacs/cider/issues/3123
- https://github.com/clojure-emacs/cider/issues/2396

**Describe the solution you'd like**

I'd like to propose a new configuration option that will instruct the interactive cider-jack-in command to prompt the user to select additional aliases to load when doing a `jack-in`. Ideally, this configuration option (name tbd) will cause cider-jack-in to examine a user's `deps.edn` or leiningen `project.clj` (or etc) for additional aliases and _explicitly prompt_ the user to select desired aliases to load in the minibuffer interactively. 

I feel this is much better UX than a prefix or global configuration that may not be applicable to all projects, or installing dir-locals and then doing more configuration. Additionally, it would be opt-in and not disruptive to users. 

**Describe alternatives you've considered**

Using `dir-locals` I do not feel is a good solution because it requires an additional emacs package. Additionally users may not wish to use the same dir-locals config every time a `jack-in` happens. Users should not have to install an additional package for this sort of thing. 

Using `cider-clojure-cli-global-options` has a similar issue as a persistant `dir-locals` config in that it will load the same alias every time which may not be always desired. (For instance, this alias may not always exist or be differently named)

Using a prefix argument to cider-jack-in with the konami code emacs chord combo of `C-u C-c C-x j j` works, but requires secret knowledge that cider-jack-in takes a prefix that allows you to specify an alias, and it is a lot to remember. 

All the above options do indeed solve the problem of loading an additional alias, but I feel strongly that it could be made much easier for users with some kind of porcelain such as an interactive prompt.

**Comment by bbatsov** [comment]
> Using dir-locals I do not feel is a good solution because it requires an additional emacs package. 

dir-locals are actually built-in Emacs functionality. No extra packages required.

In general I get your point, but historically we've avoided adding options that are specific to particular tools, as that's a slippery slope. We already plan to remove even the global options, which were actually something kind of specific to Leiningen. So, my aim is to have fewer options here (basically the name of the executable and some command-line params for it). 

Anyways, I'll keep this ticket open to see what others think about potential solution. I acknowledge that the clojure-cli is pretty popular these days and perhaps it makes some sense to provide some special support for its profiles.

**Comment by matt-y** [comment]
@bbatsov Thanks for the insight. There's always room for some custom elisp to be the glue for something like this rather than it being made part of CIDER itself and adding more tool-specific bells and whistles. Appreciate the consideration.

**Comment by bbatsov** [comment]
> Would it also apply to the existing cider-clojure-cli-aliases variable? 

Nope. It was meant to be middle-ground solution that doesn't require us to change the UI of CIDER jack-in. (e.g. to have specific prompts for aliases)

> if cider-clojure-cli-aliases is nil and we're using clojure, print the following to the minibuffer during the jack-in (or connect) process:

I'd be OK with something like this. Perhaps we can also just prompt them for aliases at this point as it would be more or less the same thing in terms of complexity and better UX. 

I also acknowledge that parsing the `deps.edn` is probably better and reduces the possibility for errors, but I think it will add a lot of complexity. If someone wants to tackle this, though - be my guest. 

In general I'd be fine with anything that doesn't add additional configuration and is limitted to some form of prompting or inference of aliases when none are specified.

**Comment by bigodel** [comment]
> We already plan to remove even the global options, which were actually something kind of specific to Leiningen.

i know this might be a [specific workflow](https://xkcd.com/1172/) issue, but i'm currently using both `cider-clojure-cli-parameters` and `cider-clojure-cli-global-options`, and they serve different purposes.

i use `cider-clojure-cli-parameters` to specify aliases with common libraries that i want to have on every REPL, such as adding aliases for `clojure.data.{csv,json,xml}`, Portal, Criterium, Tufte. so i have it set to something like `"-A:data:portal:benchmark:profiler"`.

whilst on the other hand i'm working on a project where i need to set some specific `clj-opts` (which is how the manual for `clojure` calls it) -- more specifically `-J` options that'll be passed through to `java_opts` -- to connect to a proxy, and they need to be specified before `-M`. if i use `cider-clojure-cli-parameters` it always puts it after the `-M` and that simply does not work. i might be doing something wrong as well, so if you have a better way to specify `clj-opts`.
